### PR TITLE
[server] Allow setting ws-class on project level

### LINKED ISSUE #10963: Allow specification of workspace class in gitpod.yml
## Is your feature request related to a problem? Please describe
Improve usability of https://github.com/gitpod-io/gitpod/issues/8261.

## Describe the behaviour you'd like

As a user of Gitpod, I would like to specify the workspace class that should be used for the repository in gitpod.yml.

The file would like the following:
```
workspaceRequirements:
  class: g1-standard
```

or:
```
workspaceRequirements:
  class: 
    - g1-standard
    - g2-foo-bar
```

It should behave as such:
1. The first workspace class available is used (in prebuilds and regular workspaces).
2. If no workspace class is set in the .gitpod.yml or those are not available:
     a. On Prebuilds, use the default (g1-standard)
     b. On regular workspaces, use the user preference.

## Describe alternatives you've considered
Specify workspace class on a per project basis.

## Additional context
Affected components:
- gitpod-protocol
- server

Why allow the specification of a list of workspace classes? Workspace classes names and definitions differ between gitpod installations, but the .gitpod.yml is the same for all. As such, we must allow users to define multiple a choose the first one available.

Required actions:
- Update documentation
- Talk with @gtsiolis to update the workspace class preferences to include the id, and implement it

Question: Can we have different types (array and string) being processed in the `workspaceRequirements.class` field? From @WVerlaek: Possible in Go, with a slight workaround: https://github.com/go-yaml/yaml/issues/100

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
Introduces a project settings to specify the workspace class.

Screenshot 2022-11-08 at 21 02 24

## Related Issue(s)

Fixes #10963

## Release Notes

```release-note
Workspace classes can be set in the project settings
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

File: components/dashboard/src/settings/selectClass.tsx
Comment (svenefftinge) [inline components/dashboard/src/settings/selectClass.tsx:8]: changes in this file are to extract the user-preference-specific code, making the component reusable in the project settings context.

File: components/server/ee/src/container-module.ts
Comment (svenefftinge) [inline components/server/ee/src/container-module.ts:49]: WorkspaceStarterEE does no longer container any logic, so I deleted it and also removed the corresponding bindings.

File: components/server/ee/src/prebuilds/prebuild-manager.ts
Comment (svenefftinge) [inline components/server/ee/src/prebuilds/prebuild-manager.ts:262]: I'm adding the optional project to the call graph of workspace starts here, so that we fetch the project once and pass it down instead of repeatedly asking the DB.

File: components/server/src/workspace/workspace-classes.ts
Comment (svenefftinge) [inline components/server/src/workspace/workspace-classes.ts:166]: (1/3) use of the new project setting (precedence over user setting)

File: components/server/src/workspace/workspace-starter.ts
Comment (svenefftinge) [inline components/server/src/workspace/workspace-starter.ts:227]: (2/3) use of the new project setting (precedence over user setting)

File: components/server/src/workspace/workspace-starter.ts
Comment (svenefftinge) [inline components/server/src/workspace/workspace-starter.ts:235]: (3/3) use of the new project setting (precedence over user setting)

File: components/server/src/workspace/workspace-starter.spec.ts
Comment (easyCZ) [inline components/server/src/workspace/workspace-starter.spec.ts:329]: Should there be also a test which does pass in a Project?

File: components/server/src/workspace/workspace-classes.ts
Comment (easyCZ) [inline components/server/src/workspace/workspace-classes.ts:166]: Should the project preference hold over the `entitlement service` answer whether the user gets more resources?

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:2988]: Why are we not passing the project here?

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (easyCZ) [inline components/server/src/workspace/gitpod-server-impl.ts:724]: await syntax?

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (easyCZ) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:25]: This is bit of a smell. Why is it plural workspaceClasses when we're not passing in a list?

Currently we look to only have a single preference, could we go with a list here (which is initially only 1 item long) and allow for the "preferences" to be a list?

I can imagine wanting to specify a priority list of classes I want to use. This also facilitates gradual migration away from one WS class to another over time - by changing the priorities (ordering)

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:2988]: relevant: https://github.com/gitpod-io/gitpod/pull/14535/files#r1017953380

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (geropl) [reply components/gitpod-protocol/src/teams-projects-protocol.ts:25]: I think it make senses to reuse the shape and naming we already have in place. Although I agree `WorkspaceClassConfig` or similar would have been a better choice in the first place.

File: components/server/src/workspace/workspace-classes.ts
Comment (geropl) [reply components/server/src/workspace/workspace-classes.ts:166]: :dart: Yes. `entitlementService.userGetsMoreResources` is meant to guard whether we get anything besides "default".

**Comment by geropl** [comment]
Wanted to test, but noticed the build was failing.

File: components/server/src/workspace/workspace-classes.ts
Comment (svenefftinge) [reply components/server/src/workspace/workspace-classes.ts:166]: It is not used below for the user settings. so why would this be different for the project?

Comment (geropl) [reply components/server/src/workspace/workspace-classes.ts:166]: Ah, misinterpreted the current code on multiple levels. You're right, Sven.

We do not check whether users are eligible on this level, also because we need to make up- and downwards migration through users choices and eligibility possible, including other constraints during elavation prebuilds -> regular workspaces.

This code is good as is, especially in the context of this PR. :+1:

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:168]: **thought**: It's a bit confusing that preview environments contain two quite different classes from what's on Gitpod.io and the order is also reversed, but that's out of the scope of these changes, see [relevant discussion](https://gitpod.slack.com/archives/C02F19UUW6S/p1667316174759509) (internal).

**Comment by svenefftinge** [comment]
> Wanted to test, but noticed the build was failing.

Fixed and deployed it

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:164]: **praise**: Nice component reuse! 😁

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:146]: **suggestion**: What do you think of moving the workspaces settings to the top of the general project settings here now that selecting a workspace class is a bit more prominent as it affects both workspaces and prebuilds?

| BEFORE | AFTER |
|-|-|
| Screenshot 2022-11-09 at 6 16 05 PM | Screenshot 2022-11-09 at 6 15 57 PM |

File: components/dashboard/src/projects/ProjectSettings.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/ProjectSettings.tsx:164]: **thought**: Feels good to see we're working towards making Teams & Projects more useful.

File: components/dashboard/src/settings/Preferences.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Preferences.tsx:76]: **issue**: The duplicate preference in user settings and project settings now poses a question on which preference overrides which. The workspace class scope could be something worth documenting if it's unclear but the project setting should override the user setting, correct?

> **Settings scope for workspace classes**
> Your project setting overrides the user setting.

Comment (gtsiolis) [inline components/dashboard/src/settings/Preferences.tsx:76]: **suggestion**: Once this is merged, it could be worth updating the existing and soon [outdated docs](https://www.gitpod.io/docs/configure/workspaces/workspace-classes#:~:text=You%20can%20configure%20the%20workspace%20class%20that%20should%20be%20used%20for%20your%20workspaces%20in%20your%20user%20preferences.).

> You can configure the workspace class that should be used for your workspaces in your user preferences.

**Review by gtsiolis — APPROVED** [review]
UX LGTM! Thanks for the ping[[1](https://github.com/gitpod-io/gitpod/issues/10963#issuecomment-1307760324)], @svenefftinge! 🏓 

Left some non-blocking comments that can be tackled separately in follow-up issues if needed.

Also, good call on moving this to the workspace settings section.

File: components/server/ee/src/prebuilds/prebuild-manager.ts
Comment (svenefftinge) [reply components/server/ee/src/prebuilds/prebuild-manager.ts:262]: It's not only because of reducing db queries but also a principle of good software design to explicitly pass what is needed as an argument rather than fetching data internally from a global state (the DB). I.e. more functional and easier to understand when you see the signature.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:724]: Yes, I'm creating a promise here and await it below, especially after the `mayStartPromise` which takes the longest time.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (svenefftinge) [reply components/server/src/workspace/gitpod-server-impl.ts:2988]: We don't have a project context

File: components/server/ee/src/prebuilds/prebuild-manager.ts
Comment (geropl) [reply components/server/ee/src/prebuilds/prebuild-manager.ts:262]: > also a principle of good software design to explicitly pass what is needed as an argument rather than fetching data internally from a global state (the DB)

That is a bit too vague and general for my taste to count as an argument, honestly. Then we'd need to talk service/layer boundaries first IMO, and sometimes there are good arguments against it as well.

But let's not hold up this PR any longer, I think the change is good overall. If we disagree in style we should settle that outside of PRs. :lotus_position:

**Pull request merged**
